control-plane/pkg/kafka/clientpool/clientpool.go

```diff
@@ -0,0 +1,225 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package clientpool
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/IBM/sarama"
+	"go.uber.org/zap"
+
+	corev1 "k8s.io/api/core/v1"
+	"knative.dev/eventing-kafka-broker/control-plane/pkg/kafka"
+	"knative.dev/eventing-kafka-broker/control-plane/pkg/security"
+	"knative.dev/pkg/logging"
+)
+
+const (
+	capacityPerClient = 16
+)
+
+var (
+	clients *clientPool
+)
+
+func init() {
+	cache := NewLRUCache[clientKey, sarama.Client](5*time.Minute, 30*time.Minute)
+
+	clients = &clientPool{
+		CachePool:                 cache,
+		newSaramaClient:           sarama.NewClient,
+		newClusterAdminFromClient: sarama.NewClusterAdminFromClient,
+		capacityPerClient:         capacityPerClient,
+	}
+}
+
+// A comparable struct that includes the info we need to uniquely identify a kafka cluster admin
+type clientKey struct {
+	secretName       string
+	secretNamespace  string
+	bootstrapServers string
+}
+
+type clientPool struct {
+	*CachePool[clientKey, sarama.Client]
+	newSaramaClient           kafka.NewClientFunc // use this to mock the function for tests
+	newClusterAdminFromClient kafka.NewClusterAdminFromClientFunc
+	capacityPerClient         int
+}
+
+type ReturnClientFunc func()
+type GetKafkaClientFunc func(ctx context.Context, bootstrapServers []string, secret *corev1.Secret) (sarama.Client, ReturnClientFunc, error)
+type GetKafkaClusterAdminFunc func(ctx context.Context, bootstrapServers []string, secret *corev1.Secret) (sarama.ClusterAdmin, ReturnClientFunc, error)
+
+func NilReturnClientFunc() {}
+
+func (cp *clientPool) get(ctx context.Context, bootstrapServers []string, secret *corev1.Secret) (sarama.Client, ReturnClientFunc, error) {
+	// (bootstrapServers, secret) uniquely identifies a sarama client config with the options we allow users to configure
+	key := makeClusterAdminKey(bootstrapServers, secret)
+
+	logger := logging.FromContext(ctx)
+
+	logger.Debug("about to get connection from clientpool", zap.Any("key", key))
+
+	// if a corresponding connection already exists, lets use it
+	if val, returnClient, ok, err := cp.Get(ctx, key); ok && err == nil {
+		logger.Debug("successfully got a client from the clientpool")
+		// check that the value is still good, as there are still some write errors
+		ca, err := cp.newClusterAdminFromClient(val)
+		if err != nil {
+			returnClient()
+			return nil, NilReturnClientFunc, err
+		}
+		if _, err := ca.ListTopics(); err != nil {
```

Comment:
I think we might not have permissions for this operation

Reply:
Ah okay, I added this because there were still some occasional `broken pipe` errors coming through, so I picked a random method that wouldn't modify anything to verify if the connection was still valid. Should I use something else?